### DISCUSSION TITLE
Handling of multiple interfaces

### DISCUSSION THREAD

**Discussion by jeverley** [sqm-autorate/sqm-autorate#95] [Operational Development]
I'm creating this as a place to discuss thoughts on how to implement support for multiple wan interfaces.
Not a critical initial feature, but I know a lot of users dependent upon LTE/similar connections often have failover/load balanced setups.

As a starting point for discussion, my wanmonitor service started an instance per OpenWrt wan interface. These were started and stopped by OpenWrt hotplug events. The service configuration was managed at a per interface level, with an ingress device override for the purpose of specifying a non ifb qdisc (all rx/tx rates were read from the egress device).

Not necessarily a pattern we should follow, but hopefully food for initial thought 🙂

**Comment by Fail-Safe** [comment]
I am so glad you brought this up! I've had a few people ask about multi interface support and I know to the right people, it is critical.

Two immediate thoughts come to mind:

1. I personally don't have a good test-bed for testing multiple interfaces
2. I'm wondering if we can queue this up right after we finish @Lochnair's modularization effort here: https://github.com/Fail-Safe/sqm-autorate/pull/61

@dlakelan In order to keep multiple OWD values for multiple interfaces, we would need to augment our table structures in the Lindas, yeah?

**Reply by Lochnair** [reply]
> In order to keep multiple OWD values for multiple interfaces, we would need to augment our table structures in the Lindas, yeah?

Yep, we'd have to change that up to keep it all in one instance.

But is there really a reason to force a singleton if we support multiple interfaces?

**Comment by dlakelan** [comment]
Multiple processes might well be the right way to go here, all the information needs to be duplicated anyway, like pinger threads, ping times, baselines, rates, etc. The main thing is going to be making sure two processes don't mix up each other's icmp probes, since they'll be listening on raw sockets.

**Reply by Fail-Safe** [reply]
I think this should cover us from the packet standpoint:

```lua
...
local cur_process_id = posix.getpid()
if type(cur_process_id) == "table" then
    cur_process_id = cur_process_id["pid"]
end
...
-- Set a packet ID
local packet_id = cur_process_id + 32768
```

**Reply by Fail-Safe** [reply]
We would also need to scope each config file section to an interface name or some such differentiating value.

**Reply by Lochnair** [reply]
> We would also need to scope each config file section to an interface name or some such differentiating value.

Depends I guess. If we put all interface specific options in one section, then it's just an array of sections to start instances for

**Reply by jeverley** [reply]
Adding to this, I'd recommend we want to follow a pattern of defining our config and instances per OpenWrt interface as I've described in https://github.com/Fail-Safe/sqm-autorate/pull/103

**Comment by jeverley** [comment]
As we'll need to bind our sockets to send our ping packets from the correct egress device, I've added the following setsockopt locally to facilitate some quick and dirty testing of how the service handles interfaces being brought up and down under multi-wan conditions. This is set before sending each packet.

socket.setsockopt(sock, socket.SOL_SOCKET, socket.SO_BINDTODEVICE, ul_if)

My setup has two interfaces, wan (pppoe-wan) and wwan (wwan0).
Without this the ping packets fall over to the secondary wan interface following the routing table rules rather than being pinned to the correct device.

If we want to keep the service as a single process which continually runs we'll need to make sure we properly handle devices being added and removed during runtime.
For example tun devices and wwan devices (if they're reset) require the socket to be rebound if the device is added/re-added to the system.

**Reply by Lochnair** [reply]
> If we want to keep the service as a single process which continually runs we'll need to make sure we properly handle devices being added and removed during runtime.

Yeah, and I think we need to discuss whether we want to import the speed history if a hotplug event happens so we don't need to re-learn if the interface disappears for a second

**Reply by Lochnair** [reply]
I don't know what's easiest here, making the script handle the interfaces disappearing for a while, or let it shutdown and then fire it up again with the speed history when it comes back

**Reply by jeverley** [reply]
I'm currently putting together a multi interface POC build based using a single procd instance (based on develop/develop/fix-for-gh-101 as it can handle interface drops and will pick up where it left off).
I suspect we might see some efficiency savings from doing so and it avoids the necessity of orchestrating multiple instances and their histories.
Should have it up and running locally in the next day or so to validate the approach's viability.

**Comment by farnwomt** [comment]
Just wondering if there were any other thoughts on this issue as it appears the last comments were 12 months ago?

From a personal perspective I normally have at least 2 connections to the internet, sometimes as many as 5 and I balance them using mwan3, so sqm-autorate supporting more than one interface would be very useful for me.

Will probably go with duplicating the code and creating something called sqm-autorate2 etc, but it would be nice to have something that feels like less of a hack.
